File: src/project/announcement.ts
import type { LarkChannel } from '@larksuiteoapi/node-sdk';
import { log } from '../core/logger';
import { currentBranch } from './git-info';
import { updateProject, type Project } from './registry';

/**
 * The project group's **announcement** (群公告, design §3.2) — the always-on
 * banner pinned to the top of the chat (the `📢 群公告` tab), not a Pin'd
 * message. One concise line:
 *
 *   📁 <name>   ·   📂 <cwd>   ·   🌿 <branch>
 *
 * Path shows only when the project binds an existing folder (`!blank`); the
 * default `projects/<name>` path is omitted. Branch shows only for git repos.
 * Branch is detected lazily: {@link refreshBranch} rewrites the announcement
 * on message-in / run-end, but only when the branch actually changed.
 *
 * The announcement is a docx document addressed by chat_id, so we edit it via
 * the docx block API: list blocks → find the root page block → clear its
 * children → insert one text block. `revision_id: -1` means "latest", so we
 * don't track revisions across the clear+insert.
 */
const PAGE_BLOCK_TYPE = 1;
const TEXT_BLOCK_TYPE = 2;
const LATEST_REVISION = -1;

function buildAnnouncementLine(project: Project, branch: string | null): string {
  const parts = [`📁 ${project.name}`];
  if (!project.blank) parts.push(`📂 ${project.cwd}`);
  if (branch) parts.push(`🌿 ${branch}`);
  return parts.join('   ·   ');
}

/**
 * Retry transient failures. A just-created group's announcement doc isn't
 * ready immediately — `list` blocks returns a gateway 500 (code 1771002) for a
 * short window after `chat.create`. Linear backoff rides that out (and any
 * other transient 5xx). Permission/4xx errors still get retried but fail fast
 * enough (a few seconds) and surface the real error.
 */
async function withRetry<T>(fn: () => Promise<T>, attempts = 5): Promise<T> {
  let lastErr: unknown;
  for (let i = 0; i < attempts; i++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (i < attempts - 1) await new Promise((r) => setTimeout(r, 800 * (i + 1)));
    }
  }
  throw lastErr;
}

/**
 * Replace the whole announcement with a single line of text. The list → delete
 * → create sequence is wrapped as one retry unit, and each attempt re-lists so
 * it always works off the latest revision/children (idempotent on retry).
 */
async function writeAnnouncement(channel: LarkChannel, chatId: string, line: string): Promise<void> {
  const docx = channel.rawClient.docx.v1;
  await withRetry(async () => {
    const listed = await docx.chatAnnouncementBlock.list({ path: { chat_id: chatId } });
    const items = listed.data?.items ?? [];
    const page = items.find((b) => b.block_type === PAGE_BLOCK_TYPE);
    if (!page?.block_id) throw new Error('群公告缺少 page block');

    const existing = page.children?.length ?? 0;
    if (existing > 0) {
      await docx.chatAnnouncementBlockChildren.batchDelete({
        path: { chat_id: chatId, block_id: page.block_id },
        params: { revision_id: LATEST_REVISION },
        data: { start_index: 0, end_index: existing },
      });
    }
    await docx.chatAnnouncementBlockChildren.create({
      path: { chat_id: chatId, block_id: page.block_id },
      params: { revision_id: LATEST_REVISION },
      data: {
        index: 0,
        children: [{ block_type: TEXT_BLOCK_TYPE, text: { elements: [{ text_run: { content: line } }] } }],
      },
    });
  });
}

/**
 * Pin the group announcement to the chat's top banner. `action_type: "2"`
 * means "announcement" (vs "1" = a specific message), so no message_id is
 * needed. Reuses the `im:chat` scope — no extra permission. Idempotent.
 */
async function pinAnnouncement(channel: LarkChannel, chatId: string): Promise<void> {
  await withRetry(() =>
    channel.rawClient.im.v1.chatTopNotice.putTopNotice({
      path: { chat_id: chatId },
      data: { chat_top_notice: [{ action_type: '2' }] },
    }),
  );
}

/** Set the project group's announcement, pin it to the top, persist the branch. */
export async function setAnnouncement(channel: LarkChannel, project: Project): Promise<void> {
  const branch = await currentBranch(project.cwd); // null when not a git repo
  await writeAnnouncement(channel, project.chatId, buildAnnouncementLine(project, branch));
  // Pin is best-effort: the announcement content is already written, so a pin
  // failure shouldn't lose the persisted branch below — just no top banner.
  try {
    await pinAnnouncement(channel, project.chatId);
  } catch (err) {
    log.fail('project', err, { phase: 'announcement-pin' });
  }
  // Persist the real branch only; never store a placeholder.
  await updateProject(project.name, { branch: branch ?? undefined });
}

/**
 * Lazy branch detection: re-read the git branch and, if it differs from what
 * the announcement last showed, rewrite it. Cheap when unchanged (one
 * `git rev-parse`).
 */
export async function refreshBranch(channel: LarkChannel, project: Project): Promise<void> {
  const branch = await currentBranch(project.cwd); // null when not a git repo
  if ((branch ?? '—') === (project.branch ?? '—')) return;
  log.info('project', 'branch-change', { name: project.name, from: project.branch ?? '—', to: branch ?? '—' });
  // Only persist the new branch if the rewrite succeeded — otherwise leave the
  // stored branch stale so the next message/run retries (no false "current").
  try {
    await writeAnnouncement(channel, project.chatId, buildAnnouncementLine(project, branch));
    await updateProject(project.name, { branch: branch ?? undefined });
  } catch (err) {
    log.fail('project', err, { phase: 'announcement-patch' });
  }
}

File: internal/feishu/announcement.go
package feishu

// announcement.go —— 群公告写入（对齐 TS project/announcement.ts）。
// 群公告是 docx 公告块（chat.announcement.block），写入一行文本并置顶到群顶部横幅。
// 需要飞书权限：
//   - im:chat.announcement:read        （list 公告块）
//   - im:chat.announcement:write_only  （清空 + 写入公告块）
//   - im:chat.top_notice:write_only     （置顶公告，best-effort）

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/modelzen/feishu-codex-bridge/internal/core"

	larkdocx "github.com/larksuite/oapi-sdk-go/v3/service/docx/v1"
	larkim "github.com/larksuite/oapi-sdk-go/v3/service/im/v1"
)

const (
	announcementPageBlockType  = 1
	announcementTextBlockType  = 2
	announcementLatestRevision = -1
	announcementMaxRetries     = 5
)

func ptrInt(i int) *int { return &i }

// SetGroupAnnouncement 写入群公告并置顶到群顶部横幅，对齐 TS setAnnouncement。
// text 是一行公告文本（通常形如 "📁 name · 📂 cwd · 🌿 branch"）。
func (c *Channel) SetGroupAnnouncement(ctx context.Context, chatID, text string) error {
	var lastErr error
	for attempt := 0; attempt < announcementMaxRetries; attempt++ {
		if attempt > 0 {
			time.Sleep(time.Duration(800*attempt) * time.Millisecond)
		}
		err := c.writeAnnouncementOnce(ctx, chatID, text)
		if err == nil {
			if perr := c.pinAnnouncement(ctx, chatID); perr != nil {
				// 置顶 best-effort：内容已写入，pin 失败仅告警（对齐 TS）
				core.Warn(ctx, "feishu", "announcement-pin", "置顶群公告失败（可忽略）："+perr.Error())
			}
			return nil
		}
		lastErr = err
		// scope/权限类错误重试无意义，fast-fail 并给出授权提示
		if isFatalAnnouncementErr(err) {
			return fmt.Errorf("%w（请到飞书开放平台给 bot 应用授予 im:chat.announcement:read / im:chat.announcement:write_only 权限；若也要置顶还需 im:chat.top_notice:write_only）", err)
		}
	}
	return fmt.Errorf("群公告写入失败（重试 %d 次）：%w", announcementMaxRetries, lastErr)
}

// writeAnnouncementOnce 单次写入：list 公告 page block → 清空 children → 插入一行文本。
func (c *Channel) writeAnnouncementOnce(ctx context.Context, chatID, text string) error {
	cli := c.LarkClient()

	// 1. list 公告块，找 page block（block_type==1）
	listReq := larkdocx.NewListChatAnnouncementBlockReqBuilder().ChatId(chatID).Build()
	listResp, err := cli.Docx.V1.ChatAnnouncementBlock.List(ctx, listReq)
	if err != nil {
		return fmt.Errorf("docx announcement list: %w", err)
	}
	if !listResp.Success() {
		return fmt.Errorf("docx announcement list: code=%d msg=%s", listResp.Code, listResp.Msg)
	}
	var pageID string
	existing := 0
	if listResp.Data != nil {
		for _, b := range listResp.Data.Items {
			if b.BlockType != nil && *b.BlockType == announcementPageBlockType && b.BlockId != nil {
				pageID = *b.BlockId
				existing = len(b.Children)
				break
			}
		}
	}
	if pageID == "" {
		return fmt.Errorf("群公告缺少 page block")
	}

	// 2. 清空已有 children（如有）
	if existing > 0 {
		delReq := larkdocx.NewBatchDeleteChatAnnouncementBlockChildrenReqBuilder().
			ChatId(chatID).
			BlockId(pageID).
			RevisionId(announcementLatestRevision).
			Body(larkdocx.NewBatchDeleteChatAnnouncementBlockChildrenReqBodyBuilder().
				StartIndex(0).EndIndex(existing).Build()).
			Build()
		delResp, err := cli.Docx.V1.ChatAnnouncementBlockChildren.BatchDelete(ctx, delReq)
		if err != nil {
			return fmt.Errorf("docx announcement delete: %w", err)
		}
		if !delResp.Success() {
			return fmt.Errorf("docx announcement delete: code=%d msg=%s", delResp.Code, delResp.Msg)
		}
	}

	// 3. 写入一行 text block
	textBlock := &larkdocx.Block{
		BlockType: ptrInt(announcementTextBlockType),
		Text: &larkdocx.Text{
			Elements: []*larkdocx.TextElement{
				{TextRun: &larkdocx.TextRun{Content: &text}},
			},
		},
	}
	createReq := larkdocx.NewCreateChatAnnouncementBlockChildrenReqBuilder().
		ChatId(chatID).
		BlockId(pageID).
		RevisionId(announcementLatestRevision).
		Body(larkdocx.NewCreateChatAnnouncementBlockChildrenReqBodyBuilder().
			Index(0).
			Children([]*larkdocx.Block{textBlock}).
			Build()).
		Build()
	createResp, err := cli.Docx.V1.ChatAnnouncementBlockChildren.Create(ctx, createReq)
	if err != nil {
		return fmt.Errorf("docx announcement create: %w", err)
	}
	if !createResp.Success() {
		return fmt.Errorf("docx announcement create: code=%d msg=%s", createResp.Code, createResp.Msg)
	}
	return nil
}

// pinAnnouncement 把群公告置顶到群顶部横幅（action_type=2=公告）。
func (c *Channel) pinAnnouncement(ctx context.Context, chatID string) error {
	req := larkim.NewPutTopNoticeChatTopNoticeReqBuilder().
		ChatId(chatID).
		Body(larkim.NewPutTopNoticeChatTopNoticeReqBodyBuilder().
			ChatTopNotice([]*larkim.ChatTopNotice{
				larkim.NewChatTopNoticeBuilder().ActionType("2").Build(),
			}).Build()).
		Build()
	resp, err := c.LarkClient().Im.V1.ChatTopNotice.PutTopNotice(ctx, req)
	if err != nil {
		return fmt.Errorf("im chatTopNotice put: %w", err)
	}
	if !resp.Success() {
		return fmt.Errorf("im chatTopNotice put: code=%d msg=%s", resp.Code, resp.Msg)
	}
	return nil
}

// isFatalAnnouncementErr 判断是否为 scope/权限类错误（重试无意义）。
func isFatalAnnouncementErr(err error) bool {
	if err == nil {
		return false
	}
	msg := strings.ToLower(err.Error())
	for _, kw := range []string{"scope", "permission", "无权限", "未授权", "access denied", "not allowed", "forbidden"} {
		if strings.Contains(msg, kw) {
			return true
		}
	}
	return false
}
